### ISSUE TITLE
Jakarta EE 10: Support Jakarta XML Web Services 4.0 / SOAP with Attachments 3.0

### ISSUE THREAD

**Issue by jhanders34** [OpenLiberty/open-liberty#18746]
## Description of the high level feature, including any external spec links:  
Implement the Jakarta XML Web Services 4.0 and SOAP with Attachements 3.0 specifications

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## When complete & mandatory, add links to the UFO (Upcoming Feature Overview) document, FTS (Feature Test Summary), and blogs post issues(s):
- UFO: [xmlWS-4.0](https://ibm.box.com/s/9oclr2nrbbw9uuhhdfuz9duzy7toqntg)
- FTS: #24316
- Beta Blog Post (if applicable): [Beta Blog](https://github.com/OpenLiberty/open-liberty/issues/22902)
- Blog Post: https://github.com/OpenLiberty/open-liberty/issues/24615

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## List of Steps to complete or get approvals / sign-offs for Onboarding to the Liberty release (GM date)

Instructions:
- Do the actions below and mark them complete in the checklist when they are done.
- Make sure all feature readiness approvers put the appropriate tag on the epic to indicate their approval.

## **Design**
Before Development Starts or 8 weeks before Onboarding
- [x] POC Design / UFO Review Scheduled (David Chang) or N/A.
- [x] POC Design / UFO Reviewed (Feature Owner) or N/A.
- [x] Complete any follow-ons from the POC Review.
- [x] Design / UFO Approval ([Alasdair Nottingham](https://github.com/NottyCode)) or N/A.
- [ ] No Design / No UFO Approval ([Alasdair Nottingham](https://github.com/NottyCode)) or N/A.
- [x] SVT Requirements identified. (Epic owner / Feature owner with SVT focal point)
- [x] ID Requirements identified ([Documenting Open Liberty](https://github.com/OpenLiberty/open-liberty/wiki/Documenting-Open-Liberty)). (Epic owner / Feature owner with ID focal point)
- [x] Create a child task of this epic entitled "Feature Test Summary" via [this template](https://github.com/OpenLiberty/open-liberty/issues/new?assignees=&labels=Feature+Test+Summary&template=feature_test_summary.md&title=). Add the link in above.

##
##

## **Development**
When active development has begun 
- [x] Add the "In Progress" label to this issue.

## **Beta**
If your feature, or portions of it, are going to be included in a beta  
Before Onboarding the beta
- [x] Beta Fence the functionality (`kind=beta`, `ibm:beta`, `ProductInfo.getBetaEdition()`)  

1 week before beta GA
- [x] Create, populate, and link to the [Beta blog post issue](https://github.com/OpenLiberty/open-liberty/issues/new?assignees=lauracowen%2C+jakub-pomykala&labels=&template=blog_post_beta.md&title=BETA+BLOG+-+title_of_your_update)

## **Legal**
3 weeks before Onboarding
- [x] Identify all open source libraries that are changing or are new. Work with Legal Release Services (Cass Tucker or Release PM) to get open source cleared and approved. Or N/A. (Epic Owner).   New or changed open source impacts license and Certificate of Originality.

## **Translation**
3 weeks before Onboarding
- [x] All new or changed PII messages are checked into the integration branch, before the last translation shipment out. (Epic Owner)

## **Feature Complete**
2 weeks before Onboarding
- [x] Implementation complete. (Epic owner / Feature owner)
- [x] All function tests complete. Ready for FAT Approval. (Epic owner / Feature owner)
- [ ] Review all known issues for Stop Ship. (Epic owner / Feature owner / PM)

## **Focal Point Approvals**
2 to 1 week before Onboarding
#### You **MUST** have the Design Approved or No Design Approved label before requesting focal point approvals.

All features (both "Design Approved" and "No Design Approved")
- [x] **FAT** - ([OpenLiberty/fat-approvers](https://github.com/orgs/OpenLiberty/teams/fat-approvers)). SOE FATS are running successfully or N/A . Approver adds label focalApproved:fat to the Epic in Github.
- [ ] **Demo** - ([Tom Evans](https://github.com/tevans78) or [Chuck Bridgham](https://github.com/cbridgha)). Demo is scheduled for an upcoming EOI. Approver adds label focalApproved:demo to the Epic in Github.
- [x] **Globalization** ([Sam Wong](https://github.com/samwatibm) - Liberty / Simy Cheeran - tWAS). Translation is complete or N/A. TVT - complete or N/A. Approver adds label focalApproved:globalization to the Epic in Github.

"Design Approved" features
- [x] **Accessibility** - ([Steven Zvonek](https://github.com/steven1046)). Accessibility testing is complete or N/A. Approver adds label focalApproved:accessibility to the Epic in Github.
- [x] **ID** - ([Karen Deen](https://github.com/chirp1)). Documentation work is complete or N/A . Approver adds label focalApproved:id to the Epic in Github.
- [x] **Performance** - ([Jared Anderson](https://github.com/jhanders34)). Performance testing is complete with no high severity defects or N/A . Approver adds label focalApproved:performance to the Epic in Github.
- [ ] **Serviceability** - ([Don Bourne](https://github.com/donbourne)). Serviceability has been addressed.
- [] **STE**
- [x] **SVT** - ([Brian Hanczaryk](https://github.com/hanczaryk) - APS). SVT is complete or N/A . Approver adds label focalApproved:svt to the Epic in Github.

## **Ready for GA**
1 week before Onboarding
- [ ] No Stop Ship issues for the feature. (Epic owner / Feature owner / Release PM)
- [ ] Ship Readiness Review and Release Notes completed (Epic owner / Feature owner / Release PM)
- [ ] Github Epic and Epic's issues are closed / complete. All PRs are committed to the release branch. (Epic owner / Feature owner / Backlog Subtribe PM)

1 week before GA
- [ ] Create, populate, and link to the [Blog post issue](https://github.com/OpenLiberty/open-liberty/issues/new?assignees=lauracowen%2C+jakub-pomykala&labels=&template=blog_post_ga_release.md&title=GA+BLOG+-+title_of_your_update)

## **Other deliverbles**
- [ ] **OL Guides** - ([Yee-Kang Chang](https://github.com/yeekangc)). Assessment for OL Guides is complete or N/A.
- [ ] **WDT** - (Leonard Theivendra). WDT work complete or N/A.
- [ ] **Blog** - ([Laura Cowen](https://github.com/lauracowen)) Blog article writeup (Epic owner / Feature owner / Laura Cowen)

**Comment by kwsutter** [comment]
Note that the Jakarta Web Services Metadata Specification will also be updated for [Jakarta EE 10](https://eclipse-ee4j.github.io/jakartaee-platform/jakartaee10/JakartaEE10ReleasePlan), but it will be merged with the Jakarta XML Web Services Specification.  Just wanted the team to be aware of this merging to account for the potential additional work.

**Comment by jhanders34** [comment]
Notes from Part 1 of the UFO socialization of this feature:

**Title page**
- Add SOAP with Attachments 3.0 to the title page.

**Technical Background**
- It is wsAtomicTransaction-1.2 NOT wsAtomicTransactions-1.2

**As-is / To-be**
- Remove duplicate slides (23 / 24?)

**CXF 5.0**
- Put in a statement saying that this for the future and not planned for initial xmlWS-4.0 implementation.

**Feature Design - Global Handler**
- Update E10 to be EE10
- Consider removing global handler completely instead of breaking it out into a protected feature.  
- Ask on #was-embedders and #was-liberty slack channels about whether stack products make use of this function 
- Include information on the alternatives if we decide to remove the function.

**Feature Design - wsSecurity-1.1**
- `Enabled by default` should be changed to be `Enabled by default when using xmlWS-4.0`

**State of SAX/StAX**
- XLXP is NOT in Semeru.  Need to update that.  xmlWS-4.0 will NOT use XLXP
- Do not give the option to the user to configure it.  Just use Woodstox when using xmlWS-4.0

**Feature Design**
- It is io.openliberty.jakarta.xmlWS.4.0 instead of io.openliberty.jakarta.xmlWs.4.0

**Communciation**
- Remove the first bullet
- Add a bullet about posting to openliberty.io blog post when moving to beta and ga.

**Java APIs/SPI xmlWS-4.0**
- Need to research more into the Provider API for changing XML WS 4.0 providers and the impact to our implementation. For part 2, you should have additional details for this part of the design.


Notes from Part 2 of the UFO socialization of this feature:

**Java APIs/SPIs - xmlWS-4.0**
- Update JVM properties to jaxm.properties?

**Beta**
- `read` should be `ready`

**Automated Testing**
- Add a bullet about new testing added for STAX streaming enablement.
- Add a bullet about tests for the Provider system property to switch implementations for JAXWS

**Performance**
- May need to update for GlobalHandler
- Update Woodstox bullet
- Update last bullet also it appears

**Migration Impact**
- Removal of SOAPElementFactory should be included in the migration tools

**Feature Design - Global Handler**
- Go back to the protected feature design and need to do some performance testing.

**Comment by WhiteCat22** [comment]
https://github.com/OpenLiberty/open-liberty/issues/22598

**Comment by malincoln** [comment]
@jhanders34 @neuwerk If FATs are complete, pls add target:ga label.  Thanks!

**Comment by neuwerk** [comment]
EOI Demo is set for the 17th January 2023

**Comment by malincoln** [comment]
@neuwerk is globalization required?
@samwatibm @donbourne @steven1046  can we get globlization, serviceability, and accessibility approvals?  Thanks

**Comment by samwatibm** [comment]
See email I forwarded.  Need confirmation from development of what English has changed since 23.0.0.1.

**Comment by chirp1** [comment]
Documentation issue for this epic: https://github.com/OpenLiberty/docs/issues/5854 Information to document received. Documentation being written. Approving this epic.

**Comment by donbourne** [comment]
**Serviceability Approval Comment** - Please answer the following questions for serviceability approval:

1.  UFO -- does the UFO identify the most likely problems customers will see and identify how the feature will enable them to diagnose and solve those problems without resorting to raising a PMR?  Have these issues been addressed in the implementation? _Yes, the xmlws-4.0 is nearly the same feature as the xmlWS-3.0, and uses the jakartaized version of jaxws-2.2, so most functionality is well tested. Two of the biggest changes to the specification and to the feature itself are the `jakarta.xml.ws.Provider` implementation look, and the removal of the Global Handler as a default SPI. The Global Handler change requires users to change their own user features to account for this change. Enabling Global Handlers through a user feature is well documented, but the error that occurs if you try to run the old version of the User Feature with xmlWS-4.0 will need to be covered in the documentation, in order to help users resolve the error if they see it._

2.  Test and Demo -- As part of the serviceability process we're asking feature teams to test and analyze common problem paths for serviceability and demo those problem paths to someone not involved in the development of the feature (eg. L2, test team, or another development team).  
a) What problem paths were tested and demonstrated? _The `jakarta.xml.ws.Provider` property change to look up an alternative implementation that the one Liberty ships. The Global Handler change from being a default API we enable with xmlWS-4.0_
b) Who did you demo to? _Adam Anderson_
c) Do the people you demo'd to agree that the serviceability of the demonstrated problem scenarios is sufficient to avoid PMRs for any problems customers are likely to encounter, or that L2 should be able to quickly address those problems without need to engage L3? _Yes_

3.  SVT -- SVT team is often the first team to try new features and often encounters problems setting up and using them.  Note that we're not expecting SVT to do full serviceability testing -- just to sign-off on the serviceability of the problem paths they encountered.
a) Who conducted SVT tests for this feature? _Brian Hanczaryk_
b) Do they agree that the serviceability of the problems they encountered is sufficient to avoid PMRs, or that L2 should be able to quickly address those problems without need to engage L3?  _Yes_

4.  Which L2 / L3 queues will handle PMRs for this feature?  _Yes_ 
Ensure they are present in the contact reference file and in the queue contact summary, and that the respective L2/L3 teams know they are supporting it.  Ask Don Bourne if you need links or more info.

5.  Does this feature add any new metrics or emit any new JSON events?  If yes, have you updated the JMX metrics reference list / Metrics reference list / JSON log events reference list in the Open Liberty docs? _No_

**Comment by tngiang73** [comment]
L2 is good with the STE slides - STE approving

**Comment by pgunapal** [comment]
@neuwerk Please answer the Serviceability questions in https://github.com/OpenLiberty/open-liberty/issues/18746#issuecomment-1444450783 for the serviceability approval.

**Comment by neuwerk** [comment]
@pgunapal I've updated Don's comment above with my answers. Thank you!

**Comment by neuwerk** [comment]
Keeping this issue open as there is some outstanding/ongoing working involving STAX streaming support in WS-Security as described in the UFO.

**Comment by malincoln** [comment]
@neuwerk any update on the outstanding work?  Thanks

**Comment by neuwerk** [comment]
I was able to commit the last changes that were missing to close this feature out last week #25996 and #26064. With these added, the epic is closable.
